src/NetEscapades.AspNetCore.SecurityHeaders.TagHelpers/AttributeHashTagHelper.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Linq;
+using System.Text;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using NetEscapades.AspNetCore.SecurityHeaders.Infrastructure;
+
+namespace NetEscapades.AspNetCore.SecurityHeaders.TagHelpers;
+
+/// <summary>
+/// Generates an hash of style, or inline script attributes
+/// </summary>
+[HtmlTargetElement("*", Attributes = AttributePrefix + "*")]
+public class AttributeHashTagHelper : TagHelper
+{
+    private const string AttributePrefix = "asp-add-";
+    private const string AttributeSuffix = "-to-csp";
+
+    private const CSPHashType DefaultHashType = CSPHashType.SHA256;
+
+    /// <summary>
+    /// Provides access to the <see cref="ViewContext"/>
+    /// </summary>
+    [ViewContext]
+    public ViewContext? ViewContext { get; set; }
+
+    /// <inheritdoc />
+    public override void Process(TagHelperContext context, TagHelperOutput output)
+    {
+        if (ViewContext is null)
+        {
+            throw new InvalidOperationException("ViewContext was null");
+        }
+
+        var cspAttributes = context
+            .AllAttributes
+            .Where(a =>
+                a.Name.StartsWith(AttributePrefix)
+                && a.Name.EndsWith(AttributeSuffix)
+                && a.Name != HashTagHelper.AttributeName);
+
+        foreach (var cspAttribute in cspAttributes)
+        {
+            ProcessAttribute(ViewContext, context, output, cspAttribute);
+        }
+    }
+
+    private void ProcessAttribute(
+        ViewContext viewContext,
+        TagHelperContext context,
+        TagHelperOutput output,
+        TagHelperAttribute cspAttribute)
+    {
+        var targetAttributeName = ParseTargetAttributeName(cspAttribute.Name);
+        var cspHashType = cspAttribute.ValueStyle != HtmlAttributeValueStyle.Minimized
+            ? ParseHashType(cspAttribute.Value.ToString())
+            : DefaultHashType;
+
+        var targetAttributeValue = context.AllAttributes[targetAttributeName];
+        if (targetAttributeValue is not null)
+        {
+            using var sha = CryptographyAlgorithms.Create(cspHashType);
+            var content = targetAttributeValue.Value.ToString();
+
+            // the hash is calculated based on unix line endings, not windows endings, so account for that
+            var unixContent = content!.Replace("\r\n", "\n");
+
+            var contentBytes = Encoding.UTF8.GetBytes(unixContent);
+            var hashedBytes = sha.ComputeHash(contentBytes);
+            var hash = Convert.ToBase64String(hashedBytes);
+
+            if (targetAttributeName == "style")
+            {
+                viewContext.HttpContext.SetStylesCSPHash(cspHashType, hash);
+            }
+            else
+            {
+                viewContext.HttpContext.SetScriptCSPHash(cspHashType, hash);
+            }
```

Comment:
I wonder if we should be explicit about which targets we should support here 🤔 I'm undecided though, so I think this is probably fine 👍

Reply:
Thanks for the review, and agree with all feedback. Though, now I'm back on holiday 😅, but back next week, so I will process it then.

Reply:
Yeah, I noticed that too, and I was also a bit on the fence for this, but I realized that we do the same in the content tag helper where we apply heuristics based on the tag. AFAIK, the `style` attribute is the only way that you can apply an implicit style. I think I checked the specs, but it's been a while, so not 100% sure anymore.

If you're ok with this, then I'd agree and let's keep it like this.